### DISCUSSION TITLE
@Singleton as a class decorator

### DISCUSSION THREAD

**Discussion by kchro3** [meadsteve/lagom#255] [Ideas]
In reference to this: https://github.com/meadsteve/lagom/commit/45f21a31bd8c109ac2f572912955dfe91d9279a9

I'm surprised that Lagom doesn't want to support @Inject as a class decorator, which is pretty standard in javax.inject and Guice.

In my opinion, this:
```
container = Container()
container[SomeService] = Singleton(SomeService)
main = container[MainClass]
...
```

is worse than:

```
@Singleton
class SomeService:
  pass

class MainClass
  def __init__(self, some_service: SomeService):
    pass
```

because it means that the person who wants to instantiate MainClass needs to know that SomeService should be a Singleton. The problem is worse if the class that should be a singleton is deeply nested in the dependencies.

**Comment by g0di** [comment]
Using such decorators is pretty handy indeed but I think it goes against the purpose of using dependency injection because you:
- pollute your code with things that only makes sense for a DI container (changing from one DIC to another will require you to modify your code in several places)
- removes your ability from choosing your component lifecycle for different scenarios (instantiating your object graph could be different when running an API, a CLI, or GUI, ...). 
- Tends to lead you to use your classes through the service locator anti-pattern

IMO, the object graph should be created in a single location: a composition root (note you can have many composition roots per app)

Considering this:
> it means that the person who wants to instantiate MainClass needs to know that SomeService should be a Singleton. The problem is worse if the class that should be a singleton is deeply nested in the dependencies

IMO, the person who wants to instantiate `MainClass` is actually a contributor to the project, he should knows what is he doing. I mean, DIC is not made for libraries. You do not expect some external people to install your application and use your classes directly.

**Comment by meadsteve** [comment]
I'm a bit late responding here but what @g0di lines up with the design goals of lagom. From the docs:
 
> The intention is that almost all of your code doesn't know about or rely on lagom

So adding a decorator to a class means that class "knows"/"is coupled to" lagom which is something I want to minimise.
